pkg/cmd/server/api/helpers.go

```diff
@@ -17,6 +17,85 @@ import (
 	cmdutil "github.com/openshift/origin/pkg/cmd/util"
 )
 
+var (
+	openShiftFeatureNames map[string]string
+)
+
+func init() {
+	openShiftFeatureNames = make(map[string]string, len(KnownOpenShiftFeatures))
+	openShiftFeatureNames["builder"] = "Builder"
+	openShiftFeatureNames["s2i"] = "S2I Builder"
+	openShiftFeatureNames["console"] = "Web Console"
+}
+
+// OpenShiftFeatureName returns name pleasant to read for given feature.
+func OpenShiftFeatureName(feature string) string {
+	lowerCased := strings.ToLower(feature)
+	if name, ok := openShiftFeatureNames[lowerCased]; ok {
+		return name
+	}
+	return feature
+}
+
+// Add extends feature list with given valid items. They are appended
+// unless already present.
+func (fl *FeatureList) Add(items ...string) error {
+	var (
+		err error
+		fs  = util.NewStringSet(*fl...)
+	)
+	for _, item := range items {
+		lowerCased := strings.ToLower(item)
+		if fs.Has(item) {
+			continue
+		}
+		if _, exists := openShiftFeatureNames[item]; !exists {
+			err = fmt.Errorf("unknown openshift feature %q", item)
+			continue
+		}
+		*fl = append(*fl, lowerCased)
+		fs.Insert(lowerCased)
+	}
+	return err
+}
+
+// Delete removes given items from feature list while keeping its original
+// order.
+func (fl *FeatureList) Delete(items ...string) {
+	toDelete := util.NewStringSet()
+	for _, item := range items {
+		toDelete.Insert(strings.ToLower(item))
+	}
+	preallock := 0
+	if len(*fl)-len(items) > 0 {
+		preallock = len(*fl) - len(items)
+	}
+	newList := make([]string, 0, preallock)
```

Comment:
It isn't worth it.  Feature flags are run once and should be optimized for
readability above all else

On Aug 4, 2015, at 8:51 AM, David Eads <notifications@github.com> wrote:

In pkg/cmd/server/api/helpers.go
<https://github.com/openshift/origin/pull/3953#discussion_r36182694>:

> +	}
> +	return err
> +}
> +
> +// Delete removes given items from feature list while keeping its original
> +// order.
> +func (fl *FeatureList) Delete(items ...string) {
> +	toDelete := util.NewStringSet()
> +	for _, item := range items {
> +		toDelete.Insert(strings.ToLower(item))
> +	}
> +	preallock := 0
> +	if len(*fl)-len(items) > 0 {
> +		preallock = len(*fl) - len(items)
> +	}
> +	newList := make([]string, 0, preallock)

I'd claim this isn't worth the ugly in this location, so I'd just use
[]string{}. @liggitt <https://github.com/liggitt> care to argue the other
way?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3953/files#r36182694>.